streams-tests/shared/src/test/scala/zio/stream/ZStreamSpec.scala

```diff
@@ -3873,6 +3873,19 @@ object ZStreamSpec extends ZIOBaseSpec {
               _      <- stream.tapSink(sink).take(3).runDrain
               result <- ref.get
             } yield assertTrue(result == 6)
```

Comment:
@varshith257  I agree with your investigation of tapsink behaviour. After further investigating the tapsink behaviour, its non-deterministic behaviour doesn't guarantee the sink to process all elements in of downstream cancellation. The test should not assume that all values have been processed by the sink, as this is not guaranteed. Instead, it should assert a condition that reflects the test's name: does not read ahead.

The better fix to this flaky test is `assertTrue(result <= 6) `ensures the processed values do not exceed the expected sum of the first three elements as in the test. This makes the test less prone to flakiness while still verifying that the stream does not read ahead excessively

```suggestion
            } yield assertTrue(result <= 6)
```

Reply:
Yes, It could be the intended behaviour of current tapSink implementation. I will change the test according to above as mentioned.